File: code/string_algorithms/src/kmp_algorithm/kmp.py
def getLPS(pattern):
    m = len(pattern)
    lps = [0] * (m + 1)
    i = 0
    j = -1
    lps[i] = j
    while i < m:
        while j >= 0 and pattern[i] != pattern[j]:
            j = lps[j]
        i += 1
        j += 1
        lps[i] = j

    return lps


text = "ABC ABCDAB ABCDABCDABDE"
pattern = "ABCDABD"
lps = getLPS(pattern)

# Matching part
i = 0
j = 0
n = len(text)
m = len(pattern)

while i < n - 1:
    while j >= 0 and text[i] != pattern[j]:
        j = lps[j]
    i += 1
    j += 1

    if j == m:
        print("Match found at index: " + str(i - j))
j = lps[j]

File: code/string_algorithms/src/kmp_algorithm/kmp.java
class KMP_Algorithm { 
    void KMPSearch(String pat, String txt) 
    { 
        int M = pat.length(); 
        int N = txt.length(); 
  
        int lps[] = new int[M]; 
        int j = 0; // index for pat[] 
  
        computeLPSArray(pat, M, lps); 
  
        int i = 0; // index for txt[] 
        while (i < N) { 
            if (pat.charAt(j) == txt.charAt(i)) { 
                j++; 
                i++; 
            } 
            if (j == M) { 
                System.out.println("Found pattern "+ "at index " + (i - j)); 
                j = lps[j - 1]; 
            } 
  
            // mismatch after j matches 
            else if (i < N && pat.charAt(j) != txt.charAt(i)) { 
                if (j != 0) 
                    j = lps[j - 1]; 
                else
                    i = i + 1; 
            } 
        } 
    } 
  
    void computeLPSArray(String pat, int M, int lps[]) 
    { 
        int len = 0; 
        int i = 1; 
        lps[0] = 0; // lps[0] is always 0 
  
        while (i < M) { 
            if (pat.charAt(i) == pat.charAt(len)) { 
                len++; 
                lps[i] = len; 
                i++; 
            } 
            else // (pat[i] != pat[len]) 
            { 
                if (len != 0) { 
                    len = lps[len - 1]; 
                } 
                else // if (len == 0) 
                { 
                    lps[i] = len; 
                    i++; 
                } 
            } 
        } 
    } 
  
    public static void main(String args[]) 
    { 
        String txt = "ABDABACDABABCABAB"; 
        String pat = "ABABCABAB"; 
        new KMP_String_Matching().KMPSearch(pat, txt); 
    }
} 
